### PR TITLE
Nc monitor

### PR DESCRIPTION
**Pull request by elvinhajizada**
Issue Number: #228 

Objective of pull request: To provide unittests for demonstrating updated API for new Monitor design

## Pull request checklist

Your PR fulfills the following requirements:
- [X] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [X] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [X] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [X] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [ ] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [ ] Feature
- [X] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

- There are unittests for current version of the Monitor class

## What is the new behavior?

- New (but failing) unittests for Monitor class are added to showcase and collect feedback for new API. Corresponding implementation of the new design is not here yet

## Does this introduce a breaking change?

- [X] Yes
- [ ] No

This PR is a draft and is not meant to merged yet, rather to collect feedback for new API of Monitor

### CODE REVIEW

File: tests/lava/proc/monitor/test_monitors.py [hunk 1]
```diff
@@ -342,6 +342,141 @@ def test_proc_params_accessible_in_proc_model(self):
         self.assertEqual(next(iter(exe.py_builders)).proc_params,
                          monitor.proc_params)
 
+    # New API unittests
+    def test_single_recorder_with_single_targer(self):
```
Comment (mathisrichter) [inline tests/lava/proc/monitor/test_monitors.py:346]: targer -> target

File: tests/lava/proc/monitor/test_monitors.py [hunk 2]
```diff
@@ -342,6 +342,141 @@ def test_proc_params_accessible_in_proc_model(self):
...
+        """Check if a single Recorder can monitor the spike output of a
+        neuron"""
```
Comment (mathisrichter) [inline tests/lava/proc/monitor/test_monitors.py:348]: missing full stop after "neuron"

File: tests/lava/proc/monitor/test_monitors.py [hunk 3]
```diff
@@ -342,6 +342,141 @@ def test_proc_params_accessible_in_proc_model(self):
...
+        shape = (1,)
+        num_steps = 6
+        neuron = LIF(shape=shape,
+                     vth=3,
+                     bias=1)
+
+        recorder = Recorder()
+        recorder.probe(neuron.s_out)
```
Comment (mathisrichter) [inline tests/lava/proc/monitor/test_monitors.py:356]: I'd still prefer 'monitor' as the method name. What do you think? I don't think we should stick to the "neuron" analogy with a probe and a cell as Lava is more general. I know Andreas likes "probe" better but maybe we can convince him.

File: tests/lava/proc/monitor/test_monitors.py (hunk 1 above)
Comment (mathisrichter) [inline tests/lava/proc/monitor/test_monitors.py:346]: The name should reflect that you are monitoring a Port. There should then be another unit test for monitoring an individual Var.

File: tests/lava/proc/monitor/test_monitors.py [hunk 4]
```diff
@@ -342,6 +342,141 @@ def test_proc_params_accessible_in_proc_model(self):
...
+
+        neuron.run(condition=RunSteps(num_steps=num_steps),
+                   run_cfg=Loihi1SimCfg())
+
+        spike_data = recorder.get_data(neuron.s_out)
+        self.assertTrue(np.all(spike_data == np.array([[0, 0, 1, 0, 0, 1]]).T))
+
+    def test_single_recorder_with_multiple_targets_from_single_proc(self):
+        """Check if a single Recorder can monitor multiples targets,
```
Comment (mathisrichter) [inline tests/lava/proc/monitor/test_monitors.py:365]: multiples -> multiple

File: tests/lava/proc/monitor/test_monitors.py [hunk 5]
```diff
@@ -342,6 +342,141 @@ def test_proc_params_accessible_in_proc_model(self):
...
+        here the spike output and voltage of a neuron"""
+        shape = (1,)
+        num_steps = 6
+        neuron = LIF(shape=shape,
+                     vth=3,
+                     bias=1)
+
+        recorder = Recorder()
+        recorder.probe([neuron.s_out, neuron.v])
```
Comment (mathisrichter) [inline tests/lava/proc/monitor/test_monitors.py:374]: I think we decided to not allow this anymore. There is an issue with specifying multiple sources and then also being able to specify `interval` and `offset` for them. Unless we are only supporting this kind of method call with multiple sources that all have the same `interval` and `offset`? If so, we should specify that here, maybe even put it in a comment.

File: tests/lava/proc/monitor/test_monitors.py [hunk 6]
```diff
@@ -342,6 +342,141 @@ def test_proc_params_accessible_in_proc_model(self):
...
+                   run_cfg=Loihi1SimCfg())
+
+        spike_data = recorder.get_data(neuron.s_out)
+        voltage_data = recorder.get_data(neuron.v)
+        self.assertTrue(np.all(spike_data == np.array([[0, 0, 1, 0, 0, 1]]).T))
+        self.assertTrue(
+            np.all(voltage_data == np.array([[1, 2, 0, 1, 2, 0]]).T))
+
+    def test_single_recorder_with_targets_from_multiple_procs(self):
+        """Check if a single Recorder can monitor multiples targets,
```
Comment (mathisrichter) [inline tests/lava/proc/monitor/test_monitors.py:386]: multiples -> multiple

**Review by mathisrichter — CHANGES_REQUESTED** [review]
Overall, we may not yet need these full end-to-end unit tests that run everything and then check the results. A more fine grained approach that we will have to do anyway is to unit test the API and check, for instance, only that the required Ports are instantiated in the Process. Later, we will have to add unit tests to the compiler, checking that it does the right thing with a mocked version of whatever it would receive from a network with a Recorder. Then do the same thing for the builder. And then, in the end, we would probably write a few end-to-end tests that are really required.
That approach also helps you write passing unit tests from the start, even though you are only adding functionality iteratively.

File: tests/lava/proc/monitor/test_monitors.py (hunk 5 above)
Comment (elvinhajizada) [reply tests/lava/proc/monitor/test_monitors.py:374]: Yes I agree we don't need allow multi-source monitoring, it complicates the implementation unnecessarily.

**Comment by elvinhajizada** [comment]
@mathisrichter I have some of fine-grained tests for old API, they just need to be adapted to the new API

**Pull request closed without merging**